store/src/java/com/zimbra/cs/service/mail/RecoverAccount.java

```diff
@@ -0,0 +1,170 @@
+ */
+
+package com.zimbra.cs.service.mail;
+
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.temporal.ChronoUnit;
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+
+import javax.mail.MessagingException;
+import javax.mail.internet.AddressException;
+import javax.mail.internet.InternetAddress;
+import javax.mail.internet.MimeMessage;
+import javax.mail.internet.MimeMultipart;
+
+import org.apache.commons.lang.RandomStringUtils;
+
+import com.zimbra.common.mime.MimeConstants;
+import com.zimbra.common.mime.shim.JavaMailInternetAddress;
+import com.zimbra.common.service.ServiceException;
+import com.zimbra.common.soap.Element;
+import com.zimbra.common.util.L10nUtil;
+import com.zimbra.common.util.L10nUtil.MsgKey;
+import com.zimbra.common.util.StringUtil;
+import com.zimbra.common.util.ZimbraLog;
+import com.zimbra.cs.account.Account;
+import com.zimbra.cs.account.Provisioning;
+import com.zimbra.cs.mailbox.Mailbox;
+import com.zimbra.cs.mailbox.MailboxManager;
+import com.zimbra.cs.mailbox.OperationContext;
+import com.zimbra.cs.service.util.JWEUtil;
+import com.zimbra.cs.service.util.JWEUtil.Constants;
+import com.zimbra.cs.util.AccountUtil;
+import com.zimbra.soap.ZimbraSoapContext;
+import com.zimbra.soap.mail.message.RecoverAccountRequest;
+import com.zimbra.soap.mail.message.RecoverAccountResponse;
+import com.zimbra.soap.mail.type.PasswordResetOperation;
+
+public final class RecoverAccount extends MailDocumentHandler {
+    public static final String LOG_OPERATION = "RecoverAccount:";
+    @Override
+    public Element handle(Element request, Map<String, Object> context) throws ServiceException {
+        ZimbraSoapContext zsc = getZimbraSoapContext(context);
+        OperationContext octxt = getOperationContext(zsc, context);
+
+        RecoverAccountRequest req = zsc.elementToJaxb(request);
+        PasswordResetOperation op = req.getOp();
+        if (op == null) {
+            ZimbraLog.account.debug("%s Invalid op received", LOG_OPERATION);
+            throw ServiceException.INVALID_REQUEST("Invalid op received", null);
+        }
+        String email = req.getEmail();
+        try {
+            InternetAddress ia = new JavaMailInternetAddress(email, true);
+            ia.validate();
+        } catch (AddressException ae) {
+            ZimbraLog.account.debug("%s Invalid email address: %s", LOG_OPERATION, email);
+            throw ServiceException.INVALID_REQUEST("invalid email address: " + email, ae);
+        }
+
+        Account user = Provisioning.getInstance().getAccountByName(email);
+        if (user.getPrefPasswordRecoveryAddressStatus() == null || !user.getPrefPasswordRecoveryAddressStatus().isVerified()) {
+            ZimbraLog.account.debug("%s Verified recovery email is not found for %s", LOG_OPERATION, email);
+            throw ServiceException.NOT_FOUND("Verified recovery email is not found");
```

Comment:
should not throw Not found exception. Better use Failure exception with message: something is not right, please contact admin.

Reply:
done